src/dd_init.F90

```diff
@@ -0,0 +1,326 @@
+module dd_init
+
+  use constants
+  use dd_header,  only: dd_type
+  use error,      only: fatal_error, warning
+  use global,     only: n_procs, n_particles, rank, mpi_err
+  use mesh,       only: bin_to_mesh_indices, mesh_indices_to_bin
+  use output,     only: write_message
+  use search,     only: binary_search
+  use string,     only: to_str
+
+#ifdef MPI
+  use mpi
+#endif
+
+  implicit none
+  public
+
+contains
+
+!===============================================================================
+! INITIALIZE_DOMAIN_DECOMP
+!===============================================================================
+
+  subroutine initialize_domain_decomp(dd)
+
+    type(dd_type), intent(inout) :: dd
+
+    integer :: d                ! neighbor bin
+    integer :: neighbor_meshbin
+    integer :: alloc_err        ! allocation error code
+    integer :: world_group
+    integer :: domain_master_group
+
+    call write_message("Initializing domain decomposition parameters...", 6)
+
+#ifdef _OPENMP
+    call fatal_error("Domain decomposition not implemented in " // &
+         "conjunction with OpenMP.")
```

Comment:
@smharper @paulromano  I tried to figure out why dd doesn't work with multi-threads. Several locations should account for threads parallel. One of those is the otf-tally, a key part of dd to distribute memory of tally data. If multi-threads transport different particles simultaneously,  the on-the-fly allocation of tally is possible to cause competition. So I made the allocation a critical section. 

Besides, I found maximum number of fission_bank size for thread can be reached in some cases, resulting in irreproducible results using different number of threads. I modified the size from "3*work/n_threads" to "5*work/n_threads".